modules/plugin/shapefile/src/main/java/org/geotools/data/shapefile/dbf/DbaseFileWriter.java

```diff
@@ -350,7 +350,7 @@ public String getFieldString(int size, String s) {
             // international characters must be accounted for so size != length.
             if (s != null) {
                 int maxBytes = Math.min(size, s.length());
-                buffer.replace(0, maxBytes, s);
+                buffer.insert(0, s, 0, maxBytes);
```

Comment:
So the old code was trying to replace the chars between 0 and maxBytes with the target string... and if the string ``s`` was longer, it would have expanded the buffer to fit it fully, e.g.:

```
    public static void main(String[] args) {
        StringBuffer sb = new StringBuffer("abcde");
        sb.replace(1, 3, "A_LONG_STRING");
        System.out.println(sb);
    }
```

prints ``aA_LONG_STRINGde``.  Using insert indeed limits the chars being added, for example the following:

```
    public static void main(String[] args) {
        StringBuffer sb = new StringBuffer("abcde");
        sb.insert(1, "A_LONG_STRING", 0, 2);
        System.out.println(sb);
    }
```

prints ``aA_bcde``.

So it works better for very long strings indeed. But you can see, it's worsening the situation for small strings and for cases there are no multybyte chars in the string, as the above ensures that the resulting string buffer is longer (the string is inserted by moving the chars to the left), while before no reduction would have been needed. 
In the end, the PR is replacing one performance issue with another.

I would suggest to rework this so that the insert is used against an empty buffer, and then:
- if the buffer lenght is smaller, pad the buffer with the emptystring, using another insert (the empty string is guarateed to have single byte chars, so it can be done in a single line)
- if the buffer lenght is greater, apply the same char by char reduction as now

Reply:
Good spotting, I hadn't thought about that.

I've gone back to using the replace, but before that checking the string length and trimming it down if it is too big.